### DISCUSSION TITLE
How to work with Fastapi

### DISCUSSION THREAD

**Discussion by yxlwfds** [jpsca/jinjax#64] [Q&A]
`@router.get('/article', summary='Article信息')
async def get_article(id: int, db: AsyncSession = Depends(get_db)):`

Jinjax will blocked when use async of fastapi?

**Comment by yxlwfds** [comment]
Layout.jinja:
`{#def title #}

<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <title>{{ title }}</title>
</head>
<body>
    111
  {{ content }}
</body>`

router:
`@app.get("/gc/components")
def get_components():
    catalog.render("Layout", title="Hello world", content="TEST")
`

Not work in fastapi!
![image](https://github.com/jpsca/jinjax/assets/2525362/09021b9d-7978-40ed-a1ad-77770b8740eb)

**Comment by yxlwfds** [comment]
you have to pass the request as part of the key-value pairs in the context for Jinja2. So, you also have to declare it in your path operation.
fastapi doc:https://fastapi.tiangolo.com/advanced/templates/

**Comment by thornewolf** [comment]
So this took me a little bit of time to get correct since I wanted to bind to the existing starlette `url_for` implementation.

Here are some useful code snippets
```
_TEMPLATES_DIR = "fastapi_site/templates"
templates = Jinja2Templates(_TEMPLATES_DIR)
catalog = Catalog(jinja_env=templates.env)
catalog.add_folder(_TEMPLATES_DIR)

def render(*args, **kwargs):
    return HTMLResponse(catalog.render(*args, **kwargs))

# Then later
@app.get("/")
def index(request: Request) -> Response:
    return render(
        "Index",
        request=request,
        title="Home",
    )

```

Index.jinja
```
{#def request, title #}
<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <title>My Simple Page</title>
</head>
<body>
    <h1>{{ title }}</h1>
    <SampleComponent request={request}/>
</body>
</html>
```

SampleComponent.jinja
```
{#def request #}
<ul>
        <li>Item 1</li>
        <li>Item 2</li>
        <li>Item 3</li>
        {{ url_for('index') }}
</ul>
```

**Comment by thomastraum** [comment]
@thornewolf, this helped me loads. Maybe worth adding to the documentation. thank you for posting it here. is there a way to avoid having to pass requests all the time I am using `url_for()` pretty much everywhere.

**Comment by knowsuchagency** [comment]
A simpler way IMO is to add the extension

```python
import jinjax
from fastapi import FastAPI, Request
from fastapi.templating import Jinja2Templates

app = FastAPI()

templates = Jinja2Templates(directory="templates")

# add this
templates.env.add_extension(jinjax.JinjaX)

catalog = jinjax.Catalog(jinja_env=templates.env)
catalog.add_folder("templates/components")


@app.get("/")
async def read_root(request: Request):
    return templates.TemplateResponse("index.html", {"request": request})
```

**Reply by mjgraham123** [reply]
JinjaX is awesome! It makes the building and reading templates way more efficient.
The suggestion from @knowsuchagency works great for keeping existing code working while sprinkling in components. Nice!
The only outstanding annoyance is that the use of url_for in FastAPI requires the request to be in context. That means that each component file needs to declare the request variable and all components need to pass the request in via an attribute.

Is there a way to get the request that's passed in via `catalog.render('Component', request=request)` available as a global variable in the Jinja environment (scoped to the current rendering in process)?

As a hint, here's an excerpt from the stack trace where the component didn't receive the request explicitly:
```
  File "/opt/nxtgen/app/templates/components/Page.jinja", line 4, in top-level template code
    {{ test_message }} this is broken -> {{ url_for('static', path='js/bootstrap.bundle.min.js') }}
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.12/site-packages/starlette/templating.py", line 125, in url_for
    request: Request = context["request"]
                       ~~~~~~~^^^^^^^^^^^
  File "/opt/venv/lib/python3.12/site-packages/jinja2/runtime.py", line 336, in __getitem__
    raise KeyError(key)
```

**Reply by mjgraham123** [reply]
@yxlwfds Is there any way to automatically have .render pass on the request and/or a dict of "persistent" kwargs to each .irender? I tried `request = kw.get("request")` at the top of irender and then `args["request"] = request` just before returning `component.render(**args)` and it passed the request through to components and subcomponents. Not sure this would be the best way to do that, but it solves the issue of "the request is commonly needed at any given time in any given template". `request` provides url mapping (url_for), access to current scope like current_user, etc. Very valuable.

**Reply by magoocas** [reply]
@mjgraham123 This is how I do it:
```python
content = catalog.render('Component', __globals={"url_for": request.url_for, **context})
```
context would be any other globals you want available
